### DISCUSSION TITLE
What to do when I need to unpack a dictionary?

### DISCUSSION THREAD

**Discussion by Kludex** [python/typing#1902] [Q&A]
I'm trying to add type hints to https://github.com/open-telemetry/opentelemetry-python-contrib, and in a lot of places we have a pattern that is something like:

```py
def to_type_check(fn, args, kwargs):
    return fn(*args, **kwargs)
```

How do I add type hints to this?

Some playgrounds:
- [mypy](https://gist.github.com/mypy-play/3f60edbb9056409c7d123084b3a89fe8)
- [pyright](https://pyright-play.net/?pythonVersion=3.12&strict=true&locale=en-us&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQCqKCRAxgNb0Aqi1AakRD0ACoKIQAygmqsAsACgFwqAF4ookOKkyAFACJhegJQKASqqg9pAkPtPGFCgCbVgsMAH140j6wAWMuw6wCgAXISk5FTUANrC9KYAuvSCaADO4UwsHHEAdKlpyVDsAO4Fmcxs7HmlBYlGUAC0AHxQpqEKUF1QINQwAK4gKFAhOgBUBfRjY7Ug6SaK8i5uaf0UEEgwwShTk1DTs-Md8t09fYPDoxNzaVMzZTcLzq5npB7A-SisOkThqDD0Ch-FAA-b0VjhNIwIRQJzAmANFqYEHHU69AZDKAARkc8hgnm81F8AQ4OlW6029B0WPoACYjPQAN56Vh6cJ6PxETl6eh6Jxs7EAXyMQA)

Also, I can't change how the code looks like, I can only add type hints. It's an old code source.

**Comment by TeamSpen210** [comment]
Unfortunately `ParamSpec` only works with exactly `*args, **kwargs` as inputs, so this isn't possible. If you restrict to only accepting positional arguments variadic generics would work. If these functions are only being used in the implementation and externally it just looks like the original function, perhaps you could just pretend the functions are aliases to users, and leave the implementation untypeable.

**Reply by Kludex** [reply]
> Unfortunately ParamSpec only works with exactly *args, **kwargs as inputs, so this isn't possible. If you restrict to only accepting positional arguments variadic generics would work. 

I'm aware of those points, but I'm asking for advice on what I would need to do to here to have it typed correctly.

If the answer is: "it's not possible". Then it seems like something is missing to Python itself, and I'd like to see if we can do something about it.

**Reply by Daverball** [reply]
That is the answer, the most accurate way to type this currently would be something along the lines of:

```python
def to_type_check[T](fn: Callable[..., T], args: Sequence[Any], kwargs: Mapping[str, Any]) -> T:
    return fn(*args, **kwargs)
```

`ParamSpec` is indeed a very incomplete solution and yet it took a year for support in mypy to land and in pytype there's still no full support to this day. So this should give you an idea how complex the feature is to implement, even with all the simplifications to keep the scope small. I don't think anybody would object to making `ParamSpec` more powerful, but it will take a tremendous time investment of figuring out all the important details and corner cases, so you end up with something that's still actually sound and useful.

`TypeVarTuple` is slightly more powerful, so the same function could be typed accurately for passing only positional arguments:

```python
def to_type_check[T, *Ts](fn: Callable[*Ts, T], args: tuple[*Ts]) -> T:
    return fn(*args)
```

**Reply by Kludex** [reply]
> ```py
> def to_type_check[T](fn: Callable[..., T], args: Sequence[Any], **kwargs: Mapping[str, Any]) -> T:
>     return fn(*args, **kwargs)
> ```

This is wrong. Are the `**` an oversight?

---

> That is the answer, [...]

I tried to be very objective in this discussion and not point to any direction. Again, if the answer is: "it's not possible". Then it seems like something is missing to Python itself, and I'd like to see if we can do something about it.

I don't think "but it will take a tremendous time investment of figuring out all the important details and corner cases" is a blocker here. I'm asking how can I proceed here, if that's the case.

**Reply by Daverball** [reply]
Indeed, the `**` were added accidentally, I edited the post.

I also don't think it's a blocker, I'm just warning you that it will not be easy. Many people have asked for extended `ParamSpec` features, but nobody has actually invested the time yet to write up a detailed proposal, ideally with a proof-of-concept implementation in one of the type checkers.

**Reply by Daverball** [reply]
As far how to proceed, here's the general process for proposing larger changes to typing spec (this will almost certainly require a PEP, the process for typing PEPs is no different from other PEPs and is documented elsewhere): https://github.com/python/typing-council?tab=readme-ov-file#decisions
